core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Minus;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.Pair;
+
+import com.google.common.collect.ImmutableList;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+/**
+ * Rule that replaces {@link Minus} operator with {@link Filter}
+ * when both inputs are from the same source with only filter conditions differing.
+ *
+ * <p>Example transformation:
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * EXCEPT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND NOT(comm = 5)
+ * </pre></blockquote>
+ */
+@Value.Enclosing
+public class MinusToFilterRule
+    extends RelRule<MinusToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an MinusToFilterRule. */
+  protected MinusToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Minus minus = call.rel(0);
+    if (minus.all || minus.getInputs().size() != 2) {
```

Comment:
I expect size cannot be != 2, since the rule would not be applied
So this is fine, but it could also be `assert getInputs().size() == 2;`

Reply:
I only considered the 2-way scenario. Changing to assert may allow non-2-way scenarios to enter the rule, which may cause other input filters to be ignored. I think that handwritten SQL generally does not have n-way scenarios, so 2-way is supported for the time being. Do I need to make any further changes?

Reply:
This test can build 3-way Minus
```
  @Test void testMinusToFilterRule2() {
    final Function<RelBuilder, RelNode> relFn = b -> b
        .scan("EMP")
        .project(b.field("MGR"), b.field("COMM"))
        .filter(b.lessThan(b.field("MGR"), b.literal(20)))
        .scan("EMP")
        .project(b.field("MGR"), b.field("COMM"))
        .filter(b.lessThan(b.field("MGR"), b.literal(10)))
        .scan("EMP")
        .project(b.field("MGR"), b.field("COMM"))
        .filter(b.lessThan(b.field("MGR"), b.literal(9)))
        .minus(false, 3)
        .build();

    relFn(relFn)
        .withRule(CoreRules.MINUS_TO_FILTER)
        .check();
  }
```

Reply:
I said that withOperandSupplier will only accept minus with two inputs. isn't that right?